src/reader/hdf5_reader.py

```diff
@@ -63,36 +63,61 @@ def next(self):
-                    progress(count, total, "Reading HDF5 Data")
-                    count += 1
-                    images = dataset[num_set * self.batch_size:(num_set + 1) * self.batch_size]
-                resized_images = []
-                with self.framework.trace_object('Resize', num_set / self.batch_size, 1):
-                    for image in images:
-                        resized_images.append(np.resize(image,(self._dimension,self._dimension)))
-                    sleep(.001)
-                yield resized_images
+                part_start, part_end = (int(total_samples_per_rank * self.my_rank),
+                                        int(total_samples_per_rank * (self.my_rank + 1)))
+                sample_index_list = list(range(part_start, part_end))
+            for sample_index in sample_index_list:
+                count += 1
+                logging.info(f"{utcnow()} num_set {sample_index} current batch_size {len(batch)}")
+                my_image = self._dataset[index]["data"][sample_index]
+                logging.debug(f"{utcnow()} shape of image {my_image.shape} self.max_dimension {self.max_dimension}")
+                my_image_resized = np.resize(my_image, (self.max_dimension, self.max_dimension))
+                logging.debug(f"{utcnow()} new shape of image {my_image_resized.shape}")
+                batch.append(my_image_resized)
+                is_last = 0 if count < total else 1
+                if is_last:
+                    self._dataset[index]["fp"].close()
+                    while len(batch) is not self.batch_size:
+                        batch.append(np.random.rand(self.max_dimension, self.max_dimension))
+                if len(batch) == self.batch_size:
+                    batch = np.array(batch)
+                    yield is_last, batch
+                    batch = []
+            self._dataset[index]["fp"].close()
+
+    def read_index(self, index):
+        file_index = math.floor(index / self.num_samples)
+        element_index = index % self.num_samples
+        if self._dataset[file_index]["data"] is None:
+            file_h5 = h5py.File(self._dataset[file_index]["file"], 'r')
+            self._dataset[file_index]["fp"] = file_h5
+            self._dataset[file_index]["data"] = file_h5['records']
+        my_image = self._dataset[file_index]["data"][:][:][element_index:element_index + 1]
+        my_image_resized = np.resize(my_image, (self.max_dimension, self.max_dimension))
+        logging.debug(f"{utcnow()} new shape of image {my_image_resized.shape}")
+        self._dataset[file_index]["fp"].close()
+        return my_image
+
+    def get_sample_len(self):
+        total_samples = 0
+        for element in self._dataset:
+            file_h5 = h5py.File(element["file"], 'r')
+            dataset = file_h5['records']
+            total_samples = total_samples + dataset.shape[0]
             file_h5.close()
-    def finalize(self):
-        pass
+        return total_samples
```

Comment:
Again, let us simply use the tfreader way to calculating the total number of samples instead of open the files. Even though for HDF5, the overhead is probably small. But still, imaging all the ranks are trying to open the files.

Reply:
fixed